File: tools/model/ollama-to-qmd.py
#!/usr/bin/env python3
"""
Ollama Embedding to QMD Vector Store
用 Ollama 生成向量，写入 qmd 的 SQLite vec0 表

用法:
    python3 ollama-to-qmd.py              # 嵌入所有未向量化的文档
    python3 ollama-to-qmd.py --force      # 强制重新嵌入所有文档
    python3 ollama-to-qmd.py --dry-run     # 只显示要处理的文档，不实际嵌入
"""

import sqlite3
import json
import os
import sys
import argparse
import hashlib
import time
from pathlib import Path
from dataclasses import dataclass

@dataclass
class Config:
    qmd_index: str = os.path.expanduser("~/.cache/qmd/index.sqlite")
    ollama_url: str = "http://localhost:11434/api/embeddings"
    ollama_model: str = "nomic-embed-text"
    embedding_dim: int = 768
    batch_size: int = 10
    max_content_length: int = 8000

config = Config()


def get_ollama_embedding(text: str, model: str = None) -> list[float]:
    """调用 Ollama API 获取 embedding"""
    import urllib.request
    import urllib.error
    
    model = model or config.ollama_model
    
    payload = {
        "model": model,
        "prompt": text[:config.max_content_length]
    }
    
    req = urllib.request.Request(
        config.ollama_url,
        data=json.dumps(payload).encode('utf-8'),
        headers={'Content-Type': 'application/json'}
    )
    
    try:
        with urllib.request.urlopen(req, timeout=60) as response:
            result = json.loads(response.read().decode('utf-8'))
            return result['embedding']
    except urllib.error.HTTPError as e:
        print(f"HTTP Error: {e.code} - {e.reason}")
        return None
    except Exception as e:
        print(f"Error: {e}")
        return None


def get_doc_hash(content: str, seq: int = 0) -> str:
    """生成文档哈希 (hash_seq)"""
    raw = f"{content[:1000]}{seq}"
    return hashlib.sha256(raw.encode()).hexdigest()[:64]


def get_all_docs_needing_embedding(db_path: str, force: bool = False) -> list:
    """获取需要向量化的文档"""
    conn = sqlite3.connect(db_path)
    conn.row_factory = sqlite3.Row
    cursor = conn.cursor()
    
    if force:
        cursor.execute("""
            SELECT d.path, c.doc, d.hash
            FROM documents d
            JOIN content c ON d.hash = c.hash
            WHERE d.active = 1
        """)
    else:
        cursor.execute("""
            SELECT d.path, c.doc, d.hash
            FROM documents d
            JOIN content c ON d.hash = c.hash
            LEFT JOIN content_vectors cv ON d.hash = cv.hash
            WHERE d.active = 1 AND cv.hash IS NULL
        """)
    
    results = cursor.fetchall()
    conn.close()
    
    return [(r['path'], r['doc'][:config.max_content_length], r['hash']) for r in results]


def check_existing_vectors(db_path: str) -> dict:
    """检查现有向量状态"""
    conn = sqlite3.connect(db_path)
    cursor = conn.cursor()
    
    cursor.execute("SELECT COUNT(*) FROM documents WHERE active = 1")
    total_docs = cursor.fetchone()[0]
    
    cursor.execute("SELECT COUNT(DISTINCT hash) FROM content_vectors")
    embedded_docs = cursor.fetchone()[0]
    
    conn.close()
    
    return {
        "total": total_docs,
        "embedded": embedded_docs,
        "pending": total_docs - embedded_docs
    }


def embed_documents(dry_run: bool = False, force: bool = False, model: str = None):
    """批量嵌入文档到 qmd"""
    
    model = model or config.ollama_model
    
    # 检查向量状态
    status = check_existing_vectors(config.qmd_index)
    print(f"📊 向量状态: {status['embedded']}/{status['total']} 已嵌入, {status['pending']} 待处理")
    
    if status['pending'] == 0 and not force:
        print("✅ 所有文档已向量完成!")
        return
    
    # 获取需要处理的文档
    docs = get_all_docs_needing_embedding(config.qmd_index, force)
    print(f"📄 将处理 {len(docs)} 个文档...")
    
    if dry_run:
        print("\n📝 Dry Run - 以下文档将被向量化:")
        for i, (path, _, _) in enumerate(docs[:10]):
            print(f"  {i+1}. {path}")
        if len(docs) > 10:
            print(f"  ... 还有 {len(docs) - 10} 个")
        return
    
    if not docs:
        print("✅ 没有需要处理的文档")
        return
    
    # 连接数据库准备写入
    conn = sqlite3.connect(config.qmd_index)
    cursor = conn.cursor()
    
    success_count = 0
    error_count = 0
    
    for i, (path, content, doc_hash) in enumerate(docs):
        print(f"🔄 [{i+1}/{len(docs)}] 处理: {path[:50]}...", end=" ", flush=True)
        
        # 生成向量
        embedding = get_ollama_embedding(content, model)
        
        if embedding is None:
            print("❌ 获取向量失败")
            error_count += 1
            time.sleep(1)
            continue
        
        # 写入 vectors_vec 表
        try:
            hash_seq = f"{doc_hash}|0"
            embedding_str = "[" + ",".join(map(str, embedding)) + "]"
            
            cursor.execute("""
                INSERT OR REPLACE INTO vectors_vec (hash_seq, embedding)
                VALUES (?, ?)
            """, (hash_seq, embedding_str))
            
            # 更新 content_vectors 表
            cursor.execute("""
                INSERT OR REPLACE INTO content_vectors (hash, seq, pos, model, embedded_at)
                VALUES (?, 0, 0, ?, datetime('now'))
            """, (doc_hash, model.replace("-", "")))
            
            success_count += 1
            print("✅")
            
        except Exception as e:
            print(f"❌ 写入失败: {e}")
            error_count += 1
        
        if (i + 1) % config.batch_size == 0:
            conn.commit()
            print(f"\n📦 批量提交 ({i+1}/{len(docs)})")
        
        time.sleep(0.1)
    
    conn.commit()
    conn.close()
    
    final_status = check_existing_vectors(config.qmd_index)
    print(f"\n🎉 完成!")
    print(f"   成功: {success_count}")
    print(f"   失败: {error_count}")
    print(f"   总计: {final_status['embedded']}/{final_status['total']} 已嵌入")


def main():
    parser = argparse.ArgumentParser(description="Ollama → QMD 向量嵌入工具")
    parser.add_argument("--force", "-f", action="store_true", help="强制重新嵌入所有文档")
    parser.add_argument("--dry-run", "-n", action="store_true", help="只显示待处理文档，不实际嵌入")
    parser.add_argument("--model", "-m", default=config.ollama_model, help=f"Ollama embedding 模型")
    
    args = parser.parse_args()
    
    print(f"🤖 使用模型: {args.model}")
    print(f"📁 QMD 索引: {config.qmd_index}")
    print()
    
    embed_documents(dry_run=args.dry_run, force=args.force, model=args.model)


if __name__ == "__main__":
    main()

File: tools/model/ollama-to-qmd.js
#!/usr/bin/env node
/**
 * Ollama Embedding to QMD Vector Store (Node.js version)
 * 使用 Ollama 生成向量，写入 qmd 的 SQLite vec 表
 * 
 * 用法:
 *   node ollama-to-qmd.js              # 嵌入所有未向量化的文档
 *   node ollama-to-qmd.js --force     # 强制重新嵌入所有文档
 *   node ollama-to-qmd.js --dry-run   # 只显示要处理的文档
 */

const Database = require('better-sqlite3');
const { load } = require('sqlite-vec');
const fs = require('fs');
const path = require('path');
const http = require('http');
const { promisify } = require('util');

// 配置
const QMD_INDEX = path.join(process.env.HOME, '.cache/qmd/index.sqlite');
const OLLAMA_MODEL = 'nomic-embed-text';
const BATCH_SIZE = 10;
const MAX_CONTENT_LENGTH = 8000;

// promisify http request
const httpPost = promisify((options, data, callback) => {
  const req = http.request(options, (res) => {
    let body = '';
    res.on('data', chunk => body += chunk);
    res.on('end', () => callback(null, body));
  });
  req.on('error', callback);
  req.write(JSON.stringify(data));
  req.end();
});

async function getOllamaEmbedding(text) {
  try {
    const result = await httpPost({
      hostname: 'localhost',
      port: 11434,
      path: '/api/embeddings',
      method: 'POST',
      headers: { 'Content-Type': 'application/json' }
    }, {
      model: OLLAMA_MODEL,
      prompt: text.slice(0, MAX_CONTENT_LENGTH)
    });
    
    const data = JSON.parse(result);
    return data.embedding;
  } catch (e) {
    console.error('Ollama error:', e.message);
    return null;
  }
}

function checkStatus(db) {
  const total = db.prepare('SELECT COUNT(DISTINCT hash) FROM documents WHERE active = 1').get()[0];
  const embedded = db.prepare('SELECT COUNT(DISTINCT hash) FROM content_vectors').get()[0];
  return { total, embedded, pending: total - embedded };
}

function getDocsNeedingEmbedding(db, force = false) {
  if (force) {
    return db.prepare(`
      SELECT d.path, c.doc, d.hash
      FROM documents d
      JOIN content c ON d.hash = c.hash
      WHERE d.active = 1
      GROUP BY d.hash
    `).all();
  }
  
  return db.prepare(`
    SELECT d.path, c.doc, d.hash
    FROM documents d
    JOIN content c ON d.hash = c.hash
    LEFT JOIN content_vectors cv ON d.hash = cv.hash
    WHERE d.active = 1 AND cv.hash IS NULL
    GROUP BY d.hash
  `).all();
}

async function embedDocuments(args) {
  console.log(`🤖 使用模型: ${OLLAMA_MODEL}`);
  console.log(`📁 QMD 索引: ${QMD_INDEX}\n`);
  
  // 打开数据库并加载 sqlite-vec
  const db = new Database(QMD_INDEX);
  await load(db);
  console.log('✅ sqlite-vec 扩展已加载\n');
  
  // 检查状态
  const status = checkStatus(db);
  console.log(`📊 向量状态: ${status.embedded}/${status.total} 已嵌入, ${status.pending} 待处理`);
  
  if (status.pending === 0 && !args.force) {
    console.log('✅ 所有文档已向量完成!');
    db.close();
    return;
  }
  
  // 获取待处理文档
  const docs = getDocsNeedingEmbedding(db, args.force);
  console.log(`📄 将处理 ${docs.length} 个文档...\n`);
  
  if (args.dryRun) {
    console.log('📝 Dry Run - 以下文档将被向量化:');
    docs.slice(0, 10).forEach((d, i) => console.log(`  ${i+1}. ${d.path}`));
    if (docs.length > 10) console.log(`  ... 还有 ${docs.length - 10} 个`);
    db.close();
    return;
  }
  
  if (!docs.length) {
    console.log('✅ 没有需要处理的文档');
    db.close();
    return;
  }
  
  // 批量嵌入
  let success = 0, error = 0;
  
  for (let i = 0; i < docs.length; i++) {
    const { path: docPath, doc, hash } = docs[i];
    process.stdout.write(`🔄 [${i+1}/${docs.length}] 处理: ${docPath.slice(0,40)}... `);
    
    const embedding = await getOllamaEmbedding(doc);
    
    if (!embedding) {
      console.log('❌ 获取向量失败');
      error++;
      await new Promise(r => setTimeout(r, 1000));
      continue;
    }
    
    try {
      const hashSeq = `${hash}|0`;
      
      // 写入 vectors_vec
      db.prepare('INSERT OR REPLACE INTO vectors_vec (hash_seq, embedding) VALUES (?, ?)').run(
        hashSeq,
        JSON.stringify(embedding)
      );
      
      // 更新 content_vectors 追踪表
      const now = new Date().toISOString();
      db.prepare('INSERT OR REPLACE INTO content_vectors (hash, seq, pos, model, embedded_at) VALUES (?, 0, 0, ?, ?)').run(
        hash,
        OLLAMA_MODEL.replace(/-/g, ''),
        now
      );
      
      success++;
      console.log('✅');
      
    } catch (e) {
      console.log(`❌ 写入失败: ${e.message}`);
      error++;
    }
    
    // 批量提交
    if ((i + 1) % BATCH_SIZE === 0) {
      console.log(`\n📦 批量提交 (${i+1}/${docs.length})`);
    }
    
    await new Promise(r => setTimeout(r, 50));
  }
  
  db.close();
  
  const finalStatus = checkStatus(db);
  console.log(`\n🎉 完成!`);
  console.log(`   成功: ${success}`);
  console.log(`   失败: ${error}`);
  console.log(`   总计: ${finalStatus.embedded}/${finalStatus.total} 已嵌入`);
}

// 解析参数
const args = {
  force: process.argv.includes('--force') || process.argv.includes('-f'),
  dryRun: process.argv.includes('--dry-run') || process.argv.includes('-n')
};

embedDocuments(args).catch(console.error);
